File: Algorithm/python_source/charpter16_slide/level2/max_area.py
# coding=utf-8
class MaxArea:
    def maxArea(self, height):
        i, j, res = 0, len(height) - 1, 0
        while i < j:
            if height[i] < height[j]:
                res = max(res, height[i] * (j - i))
                i += 1
            else:
                res = max(res, height[j] * (j - i))
                j -= 1
        return res


if __name__ == '__main__':
    nums = [1, 8, 6, 2, 5, 4, 8, 3, 7]

    maxArea = MaxArea()
    print maxArea.maxArea(nums)

File: Algorithm/Code_Materials/c_or_cpp/charpter16_slide/level2/max_area.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int maxArea(int* height, int heightSize) {
    int i = 0, j = heightSize - 1, res = 0;
    while (i < j) {
        res = std::max(res, std::max(height[i], height[j]) * (j - i));
        if (height[i] > height[j]) {
            i++;
        } else {
            j--;
        }
    }
    return res;
}

int main() {
    int height[] = {3, 2, 1, 4, 5, 6};
    int heightSize = sizeof(height) / sizeof(height[0]);
    std::cout << maxArea(height, heightSize) << std::endl; // 输出6，即最大子数组的长度为3，高度为6的子数组
    return 0;
}